Issue-11: Allow denormalizing of encoded JSON in Strawberryfields

marlo-longley opened the pull request:
This addresses #11 

### What is new?

Not much code. It took me some studying to understand this system in Drupal. The best article I read was [here](https://www.dx-experts.nl/en/blog/2017/drupal-8-hal-json-rest-mechanisms) even though it is specifically about hal json.

Now we can accept expanded JSON content in an SBF via POST, and create a valid new Node bearing the SBF. POST can accept either stringified or expanded JSON for SBF.


#### Testing Steps.

This section will look really long, because of pasted-in JSON. But it is not actually scary.

1. Clear cache because we have new services. Then go to `/admin/config/services/rest`. Make sure POST is enabled for node and enable the format `hal_json`. 

2. Below is a big old JSON taken from our minio storage. I was encountering permission problems creating new nodes, for the purposes of this Issue I simplified the object and removed everything on the node except for `title` and `field_descriptive_metadata`. One thing to note is the `_links` field. This is required by hal_json format. hal_json is the format I think we will be using because it can resolve relations between Drupal entities like files and images with uuids. So. If you want to do a big huge `curl` you use this (replace username:password, and token. You can get token from `http://localhost:8001/session/token`):

```

curl --include \
--request POST \
--user <username>:<password> \
--header 'Content-type: application/hal+json' \
--header 'X-CSRF-Token: <token>' \
http://localhost:8001/node\?_format\=hal_json \
--data-binary '{
"_links": {
"type": {
"href": "http://localhost:8001/rest/type/node/digital_object"
}
},
"title":[{"value":"REST ADO - test expanded JSON"}],
"field_descriptive_metadata": [
{
"value": {
"type": "Photograph",
"label": "Base ADO",
"owner": "Marlo",
"audios": null,
"images": [
30
],
"models": null,
"videos": null,
"hotspot": [],
"as:image": {
"urn:uuid:2b4fd124-ae6f-48c0-a7dd-2fd85a1b6cd2": {
"url": "s3://935/image-kite-1-2b4fd124-ae6f-48c0-a7dd-2fd85a1b6cd2.jpg",
"name": "kite_1.jpg",
"tags": [],
"type": "Image",
"dr:fid": 30,
"dr:for": "images",
"dr:uuid": "2b4fd124-ae6f-48c0-a7dd-2fd85a1b6cd2",
"checksum": "9358de99f051ae4ea44493b95f17e23b",
"crypHashFunc": "md5"
}
},
"documents": null,
"edm_agent": [
{
"name_uri": "http://example.com/marlo",
"role_uri": "",
"name_label": "Marlo",
"role_label": ""
}
],
"nominatim": {
"lat": "-102.4107493",
"lng": "34.2331373",
"city": "Earth",
"state": "Texas",
"value": "Earth, Lamb County, Texas, United States of America",
"county": "Lamb County",
"osm_id": "6595891",
"country": "United States of America",
"category": "place",
"locality": "",
"osm_type": "relation",
"postcode": "",
"country_code": "us",
"display_name": "Earth, Lamb County, Texas, United States of America",
"neighbourhood": "",
"state_district": ""
},
"ismemberof": null,
"description": "A photograph of a cool old-timey kite.",
"subject_loc": [
{
"uri": "http://id.loc.gov/authorities/subjects/sh85072580",
"label": "Kites"
}
],
"as:generator": {
"type": "Create",
"actor": {
"url": "http://localhost:8001/form/descriptive-metadata",
"name": "descriptive_metadata",
"type": "Service"
},
"endTime": "2020-01-26T22:30:23-05:00",
"summary": "Generator",
"@context": "https://www.w3.org/ns/activitystreams"
},
"date_published": "2020-01-26",
"term_aat_getty": {
"uri": "http://vocab.getty.edu/aat/300218777",
"label": "kites (equipment)"
},
"ap:entitymapping": {
"entity:file": [
"images",
"documents",
"audios",
"videos",
"models"
]
},
"geographic_place": "Earth",
"local_identifier": "kite001",
"subject_wikidata": [
{
"uri": "http://www.wikidata.org/entity/Q42861",
"label": "kite "
}
],
"strawberry_field_widget_id": "descriptive_metadata",
"strawberry_field_widget_source_entity_id": null,
"strawberry_field_widget_source_entity_uuid": ""
},
"str_flatten_keys": [
"type",
"label",
"owner",
"audios",
"images.[*]",
"models",
"videos",
"as:image.*.url",
"as:image.*.name",
"as:image.*.type",
"as:image.*.dr:fid",
"as:image.*.dr:for",
"as:image.*.dr:uuid",
"as:image.*.checksum",
"as:image.*.crypHashFunc",
"documents",
"edm_agent.[*].name_uri",
"edm_agent.[*].role_uri",
"edm_agent.[*].name_label",
"edm_agent.[*].role_label",
"nominatim.lat",
"nominatim.lng",
"nominatim.city",
"nominatim.state",
"nominatim.value",
"nominatim.county",
"nominatim.osm_id",
"nominatim.country",
"nominatim.category",
"nominatim.locality",
"nominatim.osm_type",
"nominatim.postcode",
"nominatim.country_code",
"nominatim.display_name",
"nominatim.neighbourhood",
"nominatim.state_district",
"ismemberof",
"description",
"subject_loc.[*].uri",
"subject_loc.[*].label",
"as:generator.type",
"as:generator.actor.url",
"as:generator.actor.name",
"as:generator.actor.type",
"as:generator.endTime",
"as:generator.summary",
"as:generator.@context",
"date_published",
"term_aat_getty.uri",
"term_aat_getty.label",
"ap:entitymapping.entity:file.[*]",
"geographic_place",
"local_identifier",
"subject_wikidata.[*].uri",
"subject_wikidata.[*].label",
"strawberry_field_widget_id",
"strawberry_field_widget_source_entity_id",
"strawberry_field_widget_source_entity_uuid"
]}]
}'
```
OK. You now have a new ADO. You can view it. You can “Edit” it in a webform and see all the field_descriptive_metadata properly filling out the form fields, including Nominatim, Linked Data, etc.

You might want to test POSTing a node where the SBF is not expanded. Here is another huge curl you can use for testing that:

```
curl --include \
--request POST \
--user <username>:<password> \
--header 'Content-type: application/hal+json' \
--header 'X-CSRF-Token: <token>' \
http://localhost:8001/node\?_format\=hal_json \
--data-binary '{
"_links": {
"type": {
"href": "http://localhost:8001/rest/type/node/digital_object"
}
},
"title":[{"value":"REST ADO - test string JSON"}],
"field_descriptive_metadata": [
{
"value": "{\"type\": \"Photograph\", \"label\": \"Base ADO\", \"owner\": \"Marlo\", \"audios\": null, \"images\": [30], \"models\": null, \"videos\": null, \"hotspot\": [], \"as:image\": {\"urn:uuid:2b4fd124-ae6f-48c0-a7dd-2fd85a1b6cd2\": {\"url\": \"s3://935/image-kite-1-2b4fd124-ae6f-48c0-a7dd-2fd85a1b6cd2.jpg\", \"name\": \"kite_1.jpg\", \"tags\": [], \"type\": \"Image\", \"dr:fid\": 30, \"dr:for\": \"images\", \"dr:uuid\": \"2b4fd124-ae6f-48c0-a7dd-2fd85a1b6cd2\", \"checksum\": \"9358de99f051ae4ea44493b95f17e23b\", \"crypHashFunc\": \"md5\"}}, \"documents\": null, \"edm_agent\": [{\"name_uri\": \"http://example.com/marlo\", \"role_uri\": \"\", \"name_label\": \"Marlo\", \"role_label\": \"\"}], \"nominatim\": {\"lat\": \"-102.4107493\", \"lng\": \"34.2331373\", \"city\": \"Earth\", \"state\": \"Texas\", \"value\": \"Earth, Lamb County, Texas, United States of America\", \"county\": \"Lamb County\", \"osm_id\": \"6595891\", \"country\": \"United States of America\", \"category\": \"place\", \"locality\": \"\", \"osm_type\": \"relation\", \"postcode\": \"\", \"country_code\": \"us\", \"display_name\": \"Earth, Lamb County, Texas, United States of America\", \"neighbourhood\": \"\", \"state_district\": \"\"}, \"ismemberof\": null, \"description\": \"A photograph of a cool old-timey kite.\", \"subject_loc\": [{\"uri\": \"http://id.loc.gov/authorities/subjects/sh85072580\", \"label\": \"Kites\"}], \"as:generator\": {\"type\": \"Create\", \"actor\": {\"url\": \"http://localhost:8001/form/descriptive-metadata\", \"name\": \"descriptive_metadata\", \"type\": \"Service\"}, \"endTime\": \"2020-01-26T22:30:23-05:00\", \"summary\": \"Generator\", \"@context\": \"https://www.w3.org/ns/activitystreams\"}, \"date_published\": \"2020-01-26\", \"term_aat_getty\": {\"uri\": \"http://vocab.getty.edu/aat/300218777\", \"label\": \"kites (equipment)\"}, \"ap:entitymapping\": {\"entity:file\": [\"images\", \"documents\", \"audios\", \"videos\", \"models\"]}, \"geographic_place\": \"Earth\", \"local_identifier\": \"kite001\", \"subject_wikidata\": [{\"uri\": \"http://www.wikidata.org/entity/Q42861\", \"label\": \"kite \"}], \"strawberry_field_widget_id\": \"descriptive_metadata\", \"strawberry_field_widget_source_entity_id\": null, \"strawberry_field_widget_source_entity_uuid\": \"\"}",
"str_flatten_keys": [
"type",
"label",
"owner",
"audios",
"images.[*]",
"models",
"videos",
"as:image.*.url",
"as:image.*.name",
"as:image.*.type",
"as:image.*.dr:fid",
"as:image.*.dr:for",
"as:image.*.dr:uuid",
"as:image.*.checksum",
"as:image.*.crypHashFunc",
"documents",
"edm_agent.[*].name_uri",
"edm_agent.[*].role_uri",
"edm_agent.[*].name_label",
"edm_agent.[*].role_label",
"nominatim.lat",
"nominatim.lng",
"nominatim.city",
"nominatim.state",
"nominatim.value",
"nominatim.county",
"nominatim.osm_id",
"nominatim.country",
"nominatim.category",
"nominatim.locality",
"nominatim.osm_type",
"nominatim.postcode",
"nominatim.country_code",
"nominatim.display_name",
"nominatim.neighbourhood",
"nominatim.state_district",
"ismemberof",
"description",
"subject_loc.[*].uri",
"subject_loc.[*].label",
"as:generator.type",
"as:generator.actor.url",
"as:generator.actor.name",
"as:generator.actor.type",
"as:generator.endTime",
"as:generator.summary",
"as:generator.@context",
"date_published",
"term_aat_getty.uri",
"term_aat_getty.label",
"ap:entitymapping.entity:file.[*]",
"geographic_place",
"local_identifier",
"subject_wikidata.[*].uri",
"subject_wikidata.[*].label",
"strawberry_field_widget_id",
"strawberry_field_widget_source_entity_id",
"strawberry_field_widget_source_entity_uuid"
]
}]
}'
```

marlo-longley commented:
@DiegoPino I pulled this against beta2.

marlo-longley commented:
@DiegoPino I reverted this branch to only include first of my 2 commits. All the `string_json` Encoder code was contained in the 2nd commit. Tested and all is working, using the curls I explained above.

Will do more testing with JSON:API tomorrow.

Some testing notes:
1. When configuring the REST settings in `/admin/config/services/rest` make sure you enable format `hal_json`. I didn't mention this above.
2. I was testing this on a different Archipelago and got this:
```
{"message":"Unprocessable Entity: validation failed.\nfield_file_drop.0.target_id: The referenced entity (file: 30) does not exist.\n"}                                                                                                  
```
Means, there is no File 30 on this new instance. To test, either change the number to match one on your system, or remove that key/value altogether from the POST. In my opinion this doesn't mean the code is broken. In terms of recreating content from scratch via the API we will have to
- use uuids, and
- make sure those files are loaded in before using these API calls to create content

Thoughts?

DiegoPino commented on src/Normalizer/StrawberryfieldFieldItemNormalizer.php:
```diff
@@ -67,19 +68,42 @@ public function denormalize($data, $class, $format = NULL, array $context = [])
+    
+    // Set each key in the field_item with its constructed value
+    // At this point all values will be encoded JSON strings
+    // Coming from minio data, keys should be 'value' and 'str_flatten_keys'
```
Question: 'str_flatten_keys' is a dynamic/calculated field property. Feeding it back makes no sense. If i POST a JSON without it, does this still work? Do i need to make  'str_flatten_keys'  Internal?

marlo-longley replied:
@DiegoPino `string_flatten_keys` is not needed, the POST works fine without it.
The reason I included it was because it's in the minio dump, and thought that was the test case.

DiegoPino commented on src/Normalizer/StrawberryfieldFieldItemNormalizer.php:
```diff
@@ -67,19 +68,42 @@ public function denormalize($data, $class, $format = NULL, array $context = [])
+    // Encode individually, otherwise 'value' gets nested and breaks the SBF
+    $individualEncodedValues = [];
+    foreach ($data as $key => $value ) {
+      $isJSONString = is_string($value) && is_array(json_decode($value, true)) && (json_last_error() == JSON_ERROR_NONE);
```
Can we check here if a property is calculated + readonly ? And based on that avoid setting it?

DiegoPino replied:
isJSONString should become a ::method of JSON Helper class. I have other uses for it. I actually wrote a method in a local branch for strawberry_runners, so having something like that in a re-usable method is something i want here.

DiegoPino commented:
@marlo-longley please update your pull request comment and remove the encoded string JSON denormalizer. Since we are not committing that code anymore, the pull comment needs to have that removed to avoid confusing people in the future looking at historical pulls and commits.

marlo-longley commented:
@DiegoPino OK, I addressed your code comments and updated my PR description.
I tested the new code that checks for computed/readOnly properties. Obviously, those properties still appear in the final SBF after it is saved. But if you step through the code, you can see that those properties are not being passed out if the denormalizer.

Let me know what you think. Thanks!

DiegoPino commented:
@marlo-longley changes look pretty good :) Will take me a few hours-weekened to test, but it looks quite solid now. Thanks for addressing my comments.

marlo-longley commented:
Also, for historical reference after editing my comment:
Never wrote a string JSON denormalizer as mentioned by Diego above (anyway, the denormalizer can handle encoded or decoded JSON). 
I had written an Encoder (for a normalizer), so that you could output stringified JSON if desired. Just a cosmetic thing that didn't make it into the final PR.

DiegoPino commented:
@marlo-longley did you ever got to test JSON:API?

marlo-longley commented:
@DiegoPino Thank you for merging. Did you test it?
I was planning to do JSON API testing this morning. Let me get back to you, even though I know this is merged, it can help with documentation and next steps.

marlo-longley commented:
@DiegoPino 
I just tested with JSON:API.
Our custom Normalizer and it is not even being hit (though it is hit with the REST call). So JSON:API is not finding our Normalizer.
I read this article:
https://www.lullabot.com/articles/jsonapi-2
It seems like the custom Normalizer pattern is discouraged/deprecated for JSON:API, and they suggest to use Computed Fields instead, or a couple other solutions.
Am I understanding this correctly?

DiegoPino commented:
@marlo-longley what a bummer. I see it basically does everything at the entity level, so basically to make it happen we would need to override the node serializer completely. Kinda not what i want.
Can you give https://git.drupalcode.org/project/jsonapi_extras
A look? I see it it has some config options for fieldItem normalizers (impostor) in the services.yml ,maybe we can go that far with that extra module. I sadly have no time to dig deep into this again today but will do more research, i know JSON:API was a closed system, not extendable buy i thought it was at least respecting DataType structures proposed by Symfonfy

marlo-longley commented:
@DiegoPino I will research this. I can open a new issue about it...
